File: 0442-find-all-duplicates-in-an-array/0442-find-all-duplicates-in-an-array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        answer=[]
        self.cyclicsort(nums)

        for i in range (len(nums)):
            if(nums[i] != i + 1):
                answer.append(nums[i])
            
        

        return answer
    

    def cyclicsort(self, nums: List[int]):
        i = 0
        while(i < len(nums)):
            correct = nums[i] - 1

            if(nums[i] != nums[correct]):
                self.swap(nums, i, correct)
            else :
                i+=1
            
        
    

    def swap(self, nums: List[int],  i, j):
        temp = nums[i]
        nums[i] = nums[j]
        nums[j] = temp
    

File: 0442-find-all-duplicates-in-an-array/0442-find-all-duplicates-in-an-array.java
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> answer = new ArrayList<>();
     cyclicsort(nums);

        for(int i = 0; i < nums.length; i++){
            if(nums[i] != i+1){
                answer.add(nums[i]);
            }
        }

        return answer;
    }

    void cyclicsort(int[] nums){
        int i = 0;
        while(i < nums.length){
            int correct = nums[i] - 1;

            if(nums[i] != nums[correct]){
                swap(nums, i, correct);
            } else {
                i++;
            }
        }
    }

    void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }    
    
}